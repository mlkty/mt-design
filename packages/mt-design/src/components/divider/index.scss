.mt-divider {
  $B: mt-divider;

  --mt-divider-color: var(--mt-border-color, #eee);
  --mt-divider-width: 1px;

  &--horizontal {
    display: flex;
    margin: 24px 0;
    height: var(--mt-divider-width);
    background: var(--mt-divider-color);
  }

  &--vertical {
    position: relative;
    top: -0.04em;
    display: inline-block;
    vertical-align: middle;
    margin: 0 8px;
    width: var(--mt-divider-width);
    height: 0.9em;
    background: var(--mt-divider-color);
  }

  &--with-text {
    &::before,
    &::after {
      content: '';
      width: 50%;
      height: var(--mt-divider-width);
      background: var(--mt-divider-color);
    }

    &.#{$B}--horizontal {
      height: 0;
      align-items: center;
    }
  }

  &--left {
    &::before {
      width: 5%;
    }
    &::after {
      width: 95%;
    }
  }

  &--right {
    &::before {
      width: 95%;
    }
    &::after {
      width: 5%;
    }
  }

  // ========== text ==========
  &-text {
    padding: 0 1em;
    flex-shrink: 0;
  }
}
