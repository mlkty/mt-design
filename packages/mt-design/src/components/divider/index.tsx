import './index.scss';

import {type MergeProps, withMergeProps, c} from '@mlkty/mt-shared-utils';
import {useConfigContext} from '../config-provider';

type DividerProps = MergeProps & {
  align?: 'left' | 'center' | 'right';
  direction?: 'horizontal' | 'vertical';
  children?: React.ReactNode;
};

function Divider(props: DividerProps) {
    const {getPrefixCls} = useConfigContext();
    const {
        direction = 'horizontal',
        children,
        align = 'center',
        ...restProps
    } = props;

    const prefixCls = getPrefixCls('divider');
    const hasChildren = !!children;

    const isNotVerticalMode = hasChildren && direction !== 'vertical';

    const cls = c(prefixCls, `${prefixCls}--${direction}`, {
        [`${prefixCls}--${align}`]: isNotVerticalMode,
        [`${prefixCls}--with-text`]: isNotVerticalMode,
    });

    return withMergeProps(
        restProps,
        <div className={cls} role="separator">
            {isNotVerticalMode && (
                <span className={`${prefixCls}-text`}>{children}</span>
            )}
        </div>
    );
}

export {Divider};

export type {DividerProps};
