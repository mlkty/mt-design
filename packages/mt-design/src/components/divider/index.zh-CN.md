---
title: Divider 分割线
group: 布局组件
demo:
  cols: 2
---

# Divider 分割线

帮助内容之间的分割

<code src="./demos/base.tsx">水平分割线</code>
<code src="./demos/with-text.tsx">携带文本</code>
<code src="./demos/vertical.tsx">垂直方向</code>
