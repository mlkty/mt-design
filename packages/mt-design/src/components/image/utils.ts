
export const isValidImage = (src: string) => {
    return new Promise(resolve => {
        const image = new Image();
        image.onload = () => resolve(true);
        image.onerror = () => resolve(false);
        image.src = src;
    });
};
