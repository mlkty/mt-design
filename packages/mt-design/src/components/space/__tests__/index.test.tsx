import {Space, type SpaceProps} from '..';
import {Divider} from '../../divider';
import {customRender, screen} from '../../../../tests/utils';
import mountTest from '../../../../tests/mountTest';

type SpaceElement = HTMLElement & {style: SpaceProps['style']};

describe('Space', () => {
    mountTest(Space);

    it('render empty children', async () => {
        customRender(<Space />);
        const element = screen.getByRole('space');
        expect(element.children.length).toBe(0);
    });

    it('render size snapshot', () => {
        const {container} = customRender((
            <>
                <Space>
                    <span key="custom-key-1">1</span>
                    <span key="custom-key-2">2</span>
                    no element
                </Space>
                <Space size="middle">
                    <span>1</span>
                    <span>2</span>
                </Space>
                <Space size="large">
                    <span>1</span>
                    <span>2</span>
                </Space>
                <Space size={12}>
                    <span>1</span>
                    <span>2</span>
                </Space>
            </>
        ));
        expect(container.children).toMatchSnapshot();
    });

    it('should render size gap style', () => {
        const {rerender} = customRender(
            <Space>
                <span>1</span>
                <span>2</span>
            </Space>
        );

        const container = screen.getByRole<SpaceElement>('space');

        expect(container).toHaveStyle({'--size': '8px'});

        rerender(
            <Space size="middle">
                <span>1</span>
                <span>2</span>
            </Space>
        );
        expect(container).toHaveStyle({'--size': '16px'});

        rerender(
            <Space size="large">
                <span>1</span>
                <span>2</span>
            </Space>
        );
        expect(container).toHaveStyle({'--size': '24px'});

        rerender(
            <Space size={30}>
                <span>1</span>
                <span>2</span>
            </Space>
        );
        expect(container).toHaveStyle({'--size': '30px'});

        rerender(
            // @ts-expect-error set a unrecognized string will return 0;
            <Space size="error">
                <span>1</span>
                <span>2</span>
            </Space>
        );
        expect(container).toHaveStyle({'--size': '0px'});
    });

    it('should exact classnames', () => {
        const {rerender} = customRender(<Space />);

        const container = screen.getByRole<SpaceElement>('space');
        expect(container).toHaveClass('mt-space mt-space--horizontal mt-space--align-center', {exact: true});

        rerender(<Space block />);
        expect(container).toHaveClass('mt-space--block');

        rerender(<Space block direction="vertical" />);
        expect(container).toHaveClass('mt-space--vertical');

        rerender(<Space block direction="vertical" wrap />);
        expect(container).toHaveClass('mt-space--vertical', 'mt-space--wrap');
    });

    it('should render split by divier', () => {
        customRender(
            <Space split={<Divider direction="vertical" />}>
                <span>1</span>
                <span>2</span>
                <span>3</span>
            </Space>
        );

        const container = screen.getByRole<SpaceElement>('space');
        const splitElements = screen.getAllByRole('space-split');

        expect(splitElements.length).toBe(2);
        expect(container.children).toMatchSnapshot();
    });
});
