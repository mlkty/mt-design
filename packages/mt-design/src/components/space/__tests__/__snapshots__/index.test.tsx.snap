// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button render size snapshot 1`] = `
HTMLCollection [
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 8px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      no element
    </div>
  </div>,
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 16px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
  </div>,
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 24px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
  </div>,
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 12px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
  </div>,
]
`;

exports[`Button should render split by divier 1`] = `
HTMLCollection [
  <div
    class="mt-space-item"
  >
    <span>
      1
    </span>
  </div>,
  <span
    class="mt-space-split"
    role="space-split"
  >
    <div
      class="mt-divider mt-divider--vertical"
      role="separator"
    />
  </span>,
  <div
    class="mt-space-item"
  >
    <span>
      2
    </span>
  </div>,
  <span
    class="mt-space-split"
    role="space-split"
  >
    <div
      class="mt-divider mt-divider--vertical"
      role="separator"
    />
  </span>,
  <div
    class="mt-space-item"
  >
    <span>
      3
    </span>
  </div>,
]
`;

exports[`Space render size snapshot 1`] = `
HTMLCollection [
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 8px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      no element
    </div>
  </div>,
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 16px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
  </div>,
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 24px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
  </div>,
  <div
    class="mt-space mt-space--horizontal"
    role="space"
    style="--size: 12px;"
  >
    <div
      class="mt-space-item"
    >
      <span>
        1
      </span>
    </div>
    <div
      class="mt-space-item"
    >
      <span>
        2
      </span>
    </div>
  </div>,
]
`;

exports[`Space should render split by divier 1`] = `
HTMLCollection [
  <div
    class="mt-space-item"
  >
    <span>
      1
    </span>
  </div>,
  <span
    class="mt-space-split"
    role="space-split"
  >
    <div
      class="mt-divider mt-divider--vertical"
      role="separator"
    />
  </span>,
  <div
    class="mt-space-item"
  >
    <span>
      2
    </span>
  </div>,
  <span
    class="mt-space-split"
    role="space-split"
  >
    <div
      class="mt-divider mt-divider--vertical"
      role="separator"
    />
  </span>,
  <div
    class="mt-space-item"
  >
    <span>
      3
    </span>
  </div>,
]
`;
